records/2024-04-05.database-activity-logging.md

```diff
@@ -0,0 +1,74 @@
+
+This work is related to the most recent Security Assessment and Authorization (SA&A) performed on Notify, which identified the [AU-6 - Audit Review, Analysis, and Reporting](https://www.cyber.gc.ca/en/guidance/annex-3a-security-control-catalogue-itsg-33#a33au6) control as required.
+
+## Options
+
+There are two options for auditing database activity with an RDS Aurora PostgreSQL database cluster:
+
+1. [database activity streams](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/DBActivityStreams.Overview.html); and
+2. the [PostgreSQL audit extension (pgAudit)](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/Appendix.PostgreSQL.CommonDBATasks.pgaudit.html).
+
+### Option 1: database activity streams
+
+A database activity stream logs all queries performed by the databases that are part of an Aurora cluster. The activity stream is encrypted with a KMS Customer Managed Key (CMK) and can be integrated with other AWS services.  No database instance configuration or storage space is needed to setup and use the activity stream.
+
+The proposed solution for GC Notify would be to create an asynchronous activity stream to minimize database performance impact.  This stream would then be ingested by a Kinesis Firehose which would:
+
+1. decrypt the activity events,
+1. filter events that should not be captured; and 
+1. write activity logs to an S3 bucket for auditing and alerting.
+
+![Database activity stream architecture](./diagrams/2024-04-05.database-activity-logging/activity-stream-arch.png)
+_Figure 1: [Database activity stream architecture](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/DBActivityStreams.Overview.html#DBActivityStreams.Overview.how-they-work)_
+
+In order to respect Notify's privacy policy, the S3 activity logs would be deleted after 7 days.
+
+### Option 2: PostgreSQL audit extension (pgAudit)
+
+The pgAudit extension uses the built-in PostgresSQL logging infrastructure and extends the log messages with more detail.  Logs are written to the database instances in the cluster and then published to a CloudWatch log group for auditting and alerting.  A database cluster parameter group is used to [configure which database activity is logged](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/Appendix.PostgreSQL.CommonDBATasks.pgaudit.html#Appendix.PostgreSQL.CommonDBATasks.pgaudit.reference).
+
+The proposed auditing solution using pgAudit would be to:
+
+1. enable pgAudit using the database cluster's parameter group,
+1. install and configure the pgAudit extension, and
+1. publish audit logs to a CloudWatch log group with a 7 day expiration setting.
+
+When using pgAudit, it is important to be [mindful of which queries are being logged](https://github.com/pgaudit/pgaudit/blob/master/README.md#usage-considerations).  This is because the audit logs are stored on the database instance and can lead to disk space exhaustion and performance issues if too much data is logged.
```

Comment:
Yeah as much as I really like all the options that pgsql has such as this one, having logs separate from the database seems like a better option to avoid any unintended side-effect.